python/py-frozendict/Portfile

```diff
@@ -0,0 +1,34 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+
+PortSystem          1.0
+PortGroup           python 1.0
+PortGroup           github 1.0
+
+github.setup        Marco-Sulla python-frozendict 2.3.8 v
+
+name                py-frozendict
+version             2.3.8
+revision            0
+categories-append   devel
+license             LGPL-3
+supported_archs     noarch
```

Comment:
this appears to be compiling a C extensions so I doubt this is "noarch"

Reply:
Looking at the CI it does actually compiles something with clang ; if it is not installed then one should make sure it is also not build.

Reply:
The CI build is definitely installing the extension.
```
x ./opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/frozendict/_frozendict.cpython-310-darwin.so
```

Reply:
Indeed, I see this is being compiled. I am not sure how I managed got an archive without the shared library...

I have fixed that by removing the C code after extracting the tarball. The cleanest way would be to set the env var FROZENDICT_PURE_PY=1 (https://github.com/Marco-Sulla/python-frozendict#building) at build time, but that is not yet supported by the last published tarball. I guess this could be done that way when the next version get's published.

Reply:
Now I know why I wasn't getting the C library. I tried with the py311 subport, but apparently there is only C implementation up to 3.10. So the C library was created for the 38, 39 and 310  subports but not for 311 and 312. Mystery solved :)

Reply:
Why would we want to not install the extension where supported?

Reply:
Hmm, I thought that the installation from PyPi didn't install the extension, that's why I decided not to install it. But now I realize that PyPi does not install it for python 3.11, but it does for 3.9. So I am not sure what would be the expectation from most users, and whether the extension has been tested enough for all python versions...

Reply:
I am more inclined not to ship the extension and wait for user's feedback on the need for the extension. Maybe also when upstream starts supporting 3.12 is a good moment to add the extension, since then all the subports would behave consistently.

Reply:
Please update to 2.3.9 and enable the C extension. According to the description o. PyPI it should be enabled by default since v2.3.5 and upstream provides .whl files for all Python versions with the compiled extension. Once these changes are made we will merge.

Reply:
I have updated it to now enable the C extension. It will be installed for the py38, py39, py310 subports, but not for py311, py312, since frozendict does not provide support for those python versions.
Additionally, I have checkout the v2.3.8 tag from https://github.com/Marco-Sulla/python-frozendict and run the tests with pytest. All the tests pass!
```
 test % python3.12 -m pytest  .
================================================================ test session starts ================================================================
platform darwin -- Python 3.12.0, pytest-7.4.3, pluggy-1.3.0
rootdir: /private/var/tmp/frozendict/python-frozendict
collected 241 items                                                                                                                                 

test_frozendict.py .........................................................................................................................  [ 50%]
test_frozendict_subclass.py ................................................................................................................. [ 97%]
.......                                                                                                                                       [100%]

================================================================ 241 passed in 0.09s ================================================================
 test % python3.11 -m pytest  .
================================================================ test session starts ================================================================
platform darwin -- Python 3.11.6, pytest-7.4.3, pluggy-1.3.0
rootdir: /private/var/tmp/frozendict/python-frozendict
collected 241 items                                                                                                                                 

test_frozendict.py .........................................................................................................................  [ 50%]
test_frozendict_subclass.py ................................................................................................................. [ 97%]
.......                                                                                                                                       [100%]

================================================================ 241 passed in 0.09s ================================================================
 test % python3.10 -m pytest  .
================================================================ test session starts ================================================================
platform darwin -- Python 3.10.13, pytest-7.4.3, pluggy-1.3.0
rootdir: /private/var/tmp/frozendict/python-frozendict
plugins: anyio-3.7.1
collected 482 items                                                                                                                                 

test_frozendict.py .......................................................................................................................... [ 25%]
........................................................................................................................                      [ 50%]
test_frozendict_subclass.py ................................................................................................................. [ 73%]
...............................................................................................................................               [100%]

================================================================ 482 passed in 0.22s ================================================================
 test % python3.9 -m pytest  .
================================================================ test session starts ================================================================
platform darwin -- Python 3.9.18, pytest-7.4.3, pluggy-1.3.0
rootdir: /private/var/tmp/frozendict/python-frozendict
collected 482 items                                                                                                                                 

test_frozendict.py ...............................................................................................sss........................ [ 25%]
........................................................................................................................                      [ 50%]
test_frozendict_subclass.py ..............................................................................................sss................ [ 73%]
...............................................................................................................................               [100%]

========================================================== 476 passed, 6 skipped in 0.22s ===========================================================
 test % python3.8 -m pytest  .
================================================================ test session starts ================================================================
platform darwin -- Python 3.8.18, pytest-7.4.3, pluggy-1.3.0
rootdir: /private/var/tmp/frozendict/python-frozendict
collected 482 items                                                                                                                                 

test_frozendict.py ...............................................................................................sss........................ [ 25%]
........................................................................................................................                      [ 50%]
test_frozendict_subclass.py ..............................................................................................sss................ [ 73%]
...............................................................................................................................               [100%]

========================================================== 476 passed, 6 skipped in 0.23s ===========================================================

```

Reply:
Just for confirmation, the contents for the py38, py39, py310 subports contain the C extension but not the ones for py311, py312:

```
user@host frozendict % port contents py38-frozendict | grep .so  
  /opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/frozendict/_frozendict.cpython-38-darwin.so
user@host frozendict % port contents py39-frozendict | grep .so  
  /opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/frozendict/_frozendict.cpython-39-darwin.so
user@host frozendict % port contents py310-frozendict | grep .so
  /opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/frozendict/_frozendict.cpython-310-darwin.so
user@host frozendict % port contents py311-frozendict | grep .so
user@host frozendict % port contents py312-frozendict | grep .so
user@host frozendict % 
```

Reply:
I don't understand why this is/would be dependent on the Python version. 

If, however, this is the intended behavior then the lines
"supported_archs noarch" and
"platforms {darwin any}"
should be set conditionally only for Python 3.11+

Reply:
This is because upstream has a different implementation for each python version:
https://github.com/Marco-Sulla/python-frozendict/tree/master/src/frozendict/c_src
So far there is only implementation up to python 3.10. The build system checks whether there is a directory for the python version in question and if there isn't any then it disables automatically the C extension.